include/targets/cuda/mma_tensor_op/gmem_loader.cuh

```diff
@@ -0,0 +1,802 @@
+#pragma unroll
+                for (int b = 0; b < m_batch; b++) {
+                  f_real[(m * m_batch + b) * n_dim + n] = v_real[b];
+                  f_imag[(m * m_batch + b) * n_dim + n] = v_imag[b];
+                }
+              } else {
+#pragma unroll
+                for (int b = 0; b < m_batch; b++) {
+                  zero(f_real[(m * m_batch + b) * n_dim + n], f_imag[(m * m_batch + b) * n_dim + n]);
+                }
+              }
+            }
+          }
+        }
+
+        float block_rescale_factor = 1.0f;
+        if constexpr (rescale) {
+          if constexpr (fixed) {
+            block_rescale_factor = scale_inv > 0 ? numeric_limits<half>::max() / (scale_inv * fixedMaxValue<store_t>::value) : 1.0f;
+          } else {
+            float thread_max = 0;
+#pragma unroll
+            for (int n = 0; n < n_dim; n++) {
+#pragma unroll
+              for (int m = 0; m < m_dim; m++) {
+                thread_max = abs_max(f_real[m * n_dim + n], thread_max);
+                thread_max = abs_max(f_imag[m * n_dim + n], thread_max);
+              }
+            }
+
+            // block all-reduce thread_max
+            using block_reduce_t = cub::BlockReduce<float, 1, cub::BLOCK_REDUCE_WARP_REDUCTIONS, block_y, block_z>;
+            __shared__ typename block_reduce_t::TempStorage temp_storage;
+            float block_max = block_reduce_t(temp_storage).Reduce(thread_max, cub::Max());
+
+            __shared__ float block_max_all;
+            if (threadIdx.x + blockDim.x * (threadIdx.y + blockDim.y * threadIdx.z) == 0) {
+              if (block_max > 0.0f) {
+                block_max_all = block_max;
+              } else {
+                block_max_all = 1.0f;
+              }
+            }
+            __syncthreads();
+
+            block_rescale_factor = numeric_limits<half>::max() / block_max_all;
+          }
+        }
+
+        if constexpr (std::is_same_v<load_t, half2>) {
+#pragma unroll
+          for (int n = 0; n < n_dim; n++) {
+#pragma unroll
+            for (int m = 0; m < m_dim; m++) {
+              reg_real[m * n_dim + n] = __floats2half2_rn(f_real[m * n_dim + n].x * block_rescale_factor,
+                                                          f_real[m * n_dim + n].y * block_rescale_factor);
+              reg_imag[m * n_dim + n] = __floats2half2_rn(f_imag[m * n_dim + n].x * block_rescale_factor,
+                                                          f_imag[m * n_dim + n].y * block_rescale_factor);
+            }
+          }
+        } else {
+#pragma unroll
+          for (int n = 0; n < n_dim; n++) {
+#pragma unroll
+            for (int m = 0; m < m_dim; m++) {
+              reg_real[m * n_dim + n] = f_real[m * n_dim + n] * block_rescale_factor;
+              reg_imag[m * n_dim + n] = f_imag[m * n_dim + n] * block_rescale_factor;
+            }
+          }
+        }
+
+        return 1.0f / block_rescale_factor;
```

Comment:
I think this is another division that can be removed with precomputation of `1.0 / numeric_limits<half>::max()` on the host.

Reply:
https://github.com/lattice/quda/pull/1497/commits/53c8e31e71603e0bb70114ec4af6488a8eb5f828